src/Symfony/Component/DependencyInjection/Compiler/PriorityTaggedServiceTrait.php

```diff
@@ -0,0 +1,47 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection\Compiler;
+
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Reference;
+
+/**
+ * Trait that allows a generic method to find and sort service by priority option in the tag.
+ *
+ * @author Iltar van der Berg <kjarli@gmail.com>
+ */
+trait PriorityTaggedServiceTrait
+{
+    /**
+     * Finds all services with the given tag name and order them by their priority.
+     *
+     * @param string           $tagName
+     * @param ContainerBuilder $container
+     *
+     * @return Reference[]
+     */
+    private function findAndSortTaggedServices($tagName, ContainerBuilder $container)
+    {
+        $services = $container->findTaggedServiceIds($tagName);
+
+        $queue = new \SplPriorityQueue();
+
```

Comment:
@HeahDude 

```yml
services:
    app.listener.log_request_mail:
        class: App\Listener\LogRequest
        arguments: ["@logger.mail"]
        tags: [{name: kernel.request, priority: 100}]
    app.listener.log_request_udp:
        class: App\Listener\LogRequest
        arguments: ["@logger.udp"]
        tags: [{name: kernel.request, priority: 50}]
```
This is what would register only the first one with your example. If it was unique per service, only the first encountered of each would be logged:
```yml
services:
    app.listener.log_request_mail:
        class: App\Listener\LogRequest
        arguments: ["@logger.mail"]
        tags: [{name: kernel.request, priority: 100}, {name: kernel.request, priority: 50}]
    app.listener.log_request_udp:
        class: App\Listener\LogRequest
        arguments: ["@logger.udp"]
        tags: [{name: kernel.request, priority: 50}, {name: kernel.request, priority: 100}]
```

So by service id is feasible as this example makes no sense in the first place. However, it would also mean that the mail would be registered as 100 and the udp with 50.

Imo regarding this PR, I think I'll leave this as is. If it's a behavior that needs to be changed, a PR can be opened to deprecated it as 3.4 and 4.0 are still 1.5 year away.